csrc/welford.cu

```diff
@@ -0,0 +1,415 @@
+    space_size *= input.size(i);
+  }
+
+  int block_x = 16;
+  const dim3 block(block_x, batch_size);
+  const dim3 grid(feature_size);
+  // shared memory used for reduce;
+  int smem_size = batch_size * block_x * 2 * sizeof(int);
+
+  AT_DISPATCH_FLOATING_TYPES(input.type(), "batchnorm_forward", ([&] {
+    reduce_bn_kernel<scalar_t><<<grid, block, smem_size>>>(
+        input.data<scalar_t>(),
+        grad_output.data<scalar_t>(),
+        mean.data<scalar_t>(),
+        var.data<scalar_t>(),
+        mean_dy.data<scalar_t>(),
+        mean_dy_xmu.data<scalar_t>(),
+        grad_weight.data<scalar_t>(),
+        grad_bias.data<scalar_t>(),
+        batch_size,
+        feature_size,
+        space_size,
+        eps);
+  }));
+  
+  return {mean_dy, mean_dy_xmu, grad_weight, grad_bias};
+}
+
+at::Tensor batchnorm_backward_CUDA(
+    const at::Tensor grad_output,
+    const at::Tensor input,
+    const at::Tensor mean,
+    const at::Tensor var,
+    const at::Tensor weight,
+    const at::Tensor shift,
+    const at::Tensor mean_dy,
+    const at::Tensor mean_dy_xmu,
+    const float eps) {
+  const auto batch_size = input.size(0);
+  const auto feature_size = input.size(1);
+
+  at::Tensor grad_input = at::empty_like(input);
+
+  auto space_size = input.size(2);
+  for (int i = 3; i < input.ndimension(); i++) {
+    space_size *= input.size(i);
+  }
+
+  const dim3 block(512);
+  // TODO(jie): should I do 1 block per feature?
+  const dim3 grid(feature_size, batch_size);
+
+  AT_DISPATCH_FLOATING_TYPES(input.type(), "batchnorm_forward", ([&] {
+    batchnorm_backward_kernel<scalar_t><<<grid, block>>>(
+        grad_output.data<scalar_t>(),
+        input.data<scalar_t>(),
+        mean.data<scalar_t>(),
+        var.data<scalar_t>(),
+        weight.data<scalar_t>(),
+        mean_dy.data<scalar_t>(),
+        mean_dy_xmu.data<scalar_t>(),
+        grad_input.data<scalar_t>(),
+        space_size,
+        eps);
+  }));
+  
+  return grad_input;
+}
+
+template <typename scalar_t>
+__global__ void welford_kernel_parallel(
+      const scalar_t* __restrict__ mean,
+      const scalar_t* __restrict__ var_biased,
+      scalar_t* __restrict__ out_mean,
+      scalar_t* __restrict__ out_var,
+      scalar_t* __restrict__ out_var_biased,
+      const int ns,
+      const int fs,
+      const int numel) {
+  extern __shared__ int s_mem[];
+  int block_size = blockDim.x;
+
+  float *mean_l = (float*) s_mem;
+  float *m2n_l = (float*) &(s_mem[block_size]);
+  int *num_item_l = (int*) &(s_mem[block_size*2]);
+
+  int input_base = blockIdx.x*ns + threadIdx.x;
+  int thread_id = threadIdx.x;
+
+  // load data; 
+  mean_l[thread_id] = mean[input_base];
+  m2n_l[thread_id] = var_biased[input_base] * numel;
+  num_item_l[thread_id] = numel;
+
+  __syncthreads();
+  
+  // parallel reduce with interleaved threads
+  // TODO(jie): unroll this?
+  for (int offset = powf(2.0, ceil(log2f(block_size))); offset > 0; offset>>=1) {
```

Comment:
Instead of int offset = powf(2.0, ceil(log2f(block_size))), which is a float->int conversion, three [heavyweight](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#math-libraries) math library calls, then a float->int coversion, this gadget I wrote might be helpful:
https://github.com/pytorch/pytorch/blob/66b3bae47c61cfc6c345974c60fcd734ffe84365/aten/src/THC/THCReduce.cuh#L35-L41

I have some questions about your launch configurations though, which would have to interact with this.  Let's talk offline.